staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go

```diff
@@ -890,6 +891,97 @@ func TestEnvelopeMetrics(t *testing.T) {
 	}
 }
 
+// This test validates the correctness of the apiserver_envelope_encryption_dek_source_cache_size metric
+// 1. Multiple transformers are created, which should result in unique cache size for each provider
+// 2. A transformer with known number of states was created to encrypt, then on restart, another transformer
+// was created, which should result in expected number of cache keys for all the decryption calls for each
+// state used previously for encryption.
+func TestEnvelopeMetricsCache(t *testing.T) {
+	envelopeService := newTestEnvelopeService()
+	envelopeService.keyVersion = testKeyVersion
+	state, err := testStateFunc(testContext(t), envelopeService, clock.RealClock{}, randomBool())()
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := testContext(t)
+	dataCtx := value.DefaultContext(testContextText)
+	provider1 := "one"
+	provider2 := "two"
+	numOfStates := 10
+
+	testCases := []struct {
+		desc    string
+		metrics []string
+		want    string
+	}{
+		{
+			desc: "dek source cache size",
+			metrics: []string{
+				"apiserver_envelope_encryption_dek_source_cache_size",
+			},
+			want: fmt.Sprintf(`
+				# HELP apiserver_envelope_encryption_dek_source_cache_size [ALPHA] Number of records in data encryption key (DEK) source cache. On a restart, this value is an approximation of the number of decrypt RPC calls the server will make to the KMS plugin.
+				# TYPE apiserver_envelope_encryption_dek_source_cache_size gauge
+        		apiserver_envelope_encryption_dek_source_cache_size{provider_name="%s"} %d
+        		apiserver_envelope_encryption_dek_source_cache_size{provider_name="%s"} 1
+				`, provider1, numOfStates, provider2),
+		},
+	}
+	testStateFunc1 := func() (State, error) { return state, nil }
+	transformer1 := NewEnvelopeTransformer(envelopeService, provider1, testStateFunc1)
```

Comment:
I think you can just use `testStateFunc` here directly and then drop `transformer1.(*envelopeTransformer).stateFunc = func() {` logic below.

```suggestion
	transformer1 := NewEnvelopeTransformer(envelopeService, provider1, testStateFunc(ctx, envelopeService, clock.RealClock{}, randomBool())
```